Use parent to pick proper document during parsing

cbrewster opened the pull request:
<!-- Please describe your changes on the following line: -->
Relies on servo/html5ever#312

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18277 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18536)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/servoparser/async_html.rs, components/script/dom/servoparser/mod.rs
 * @kichjang: components/script/dom/servoparser/async_html.rs, components/script/dom/servoparser/mod.rs

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18541) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@bors-servo delegate=cynicaldevil

@cynicaldevil Would you be interested in reviewing this?

bors-servo commented:
:v: @cynicaldevil can now approve this pull request

jdm commented on components/script/dom/servoparser/async_html.rs:
```diff
@@ -335,14 +336,14 @@ impl Tokenizer {
                               .find(|attr| attr.name.local.eq_str_ignore_ascii_case("is"))
                               .map(|attr| LocalName::from(&*attr.value));
-
+                let document = parent.and_then(|p| self.nodes.get(&p)).map(|n| document_from_node(&**n)).unwrap_or_else(|| Root::from_ref(&*self.document));
```
Let's make this a bit more readable with some newlines:
```rust
let document =
    parent
    .and_then(|p| self.nodes.get(&p))
    .map(|n| document_from_node(&**n))
    .unwrap_or_else(|| Root::from_ref(&*self.document));
```

jdm commented on Cargo.toml:
```diff
@@ -19,3 +19,5 @@ opt-level = 3
 #
 #     <crate> = { path = "/path/to/local/checkout" }
+html5ever = { path = "../html5ever/html5ever" }
+xml5ever = { path = "../html5ever/xml5ever" }
```
These will need to be removed.

cbrewster replied:
Is it okay to wait for servo/html5ever#312 to merge first?

jdm reviewed (changes requested):
Sorry about the long delay! @cbrewster do you have time to rebase this and address a couple comments?

cbrewster committed:
Use the correct document when creating elements

The html parser now sends the intended parent when requesting a new
element to be created. This will ensure that elements are created using
the proper document.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21223) made this pull request unmergeable. Please resolve the merge conflicts.

cbrewster commented:
I don't have the bandwidth to finish this at the moment.
